uplink/src/collector/telemetry.rs

```diff
@@ -0,0 +1,534 @@
+        self.timestamp = timestamp;
+        self.sequence = sequence;
+    }
+}
+
+impl Point for Process {
+    fn sequence(&self) -> u32 {
+        self.sequence
+    }
+
+    fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+}
+
+impl Package for Buffer<Process> {
+    fn topic(&self) -> Arc<String> {
+        self.topic.clone()
+    }
+
+    fn serialize(&self) -> Vec<u8> {
+        serde_json::to_vec(&self.buffer).unwrap()
+    }
+
+    fn anomalies(&self) -> Option<(String, usize)> {
+        self.anomalies()
+    }
+}
+
+type Sequence = u32;
+
+/// Holds current sequence number values meant for different data streams
+#[derive(Default)]
+struct Sequences {
+    system: Sequence,
+    processors: Sequence,
+    processes: Sequence,
+    disks: Sequence,
+    networks: Sequence,
+}
+
+/// Stream handles for the various data streams
+struct Streams {
+    system: Stream<SystemStats>,
+    processors: Stream<Processor>,
+    processes: Stream<Process>,
+    disks: Stream<Disk>,
+    networks: Stream<Network>,
+}
+
+impl Streams {
+    fn init(tx: Sender<Box<dyn Package>>, config: &Arc<Config>) -> Self {
+        Streams {
+            system: Stream::dynamic(
+                "system_stats",
+                &config.project_id,
+                &config.device_id,
+                tx.clone(),
+            ),
+            processors: Stream::dynamic(
+                "processor_stats",
+                &config.project_id,
+                &config.device_id,
+                tx.clone(),
+            ),
+            processes: Stream::dynamic(
+                "process_stats",
+                &config.project_id,
+                &config.device_id,
+                tx.clone(),
+            ),
+            disks: Stream::dynamic("disk_stats", &config.project_id, &config.device_id, tx.clone()),
+            networks: Stream::dynamic("network_stats", &config.project_id, &config.device_id, tx),
+        }
+    }
+}
+
+/// Collects and forward system information such as kernel version, memory and disk space usage,
+/// information regarding running processes, network and processor usage, etc to an IoT platform.
+pub struct StatCollector {
+    /// Handle to sysinfo struct containing system information.
+    sys: System,
+    /// Frequently updated timestamp value, used to ensure idempotence.
+    timestamp: u64,
+    /// System information values to be serialized.
+    stats: SystemStats,
+    /// Information about running processes.
+    processes: HashMap<i32, Process>,
+    /// Individual Processor information.
+    processors: HashMap<String, Processor>,
+    /// Information regarding individual Network interfaces.
+    networks: HashMap<String, Network>,
+    /// Information regarding individual Disks.
+    disks: HashMap<String, Disk>,
+    /// Uplink configuration.
+    config: Arc<Config>,
+    /// Stream handles
+    streams: Streams,
+    /// Data point sequence numbers.
+    sequences: Sequences,
```

Comment:
Why can't we just update sequence number in respective structs (Process, Processor, SystemStats) instead of maintaining a new `Sequences` struct, incrementing it and passing it to `update` methods?

Reply:
The sequence numbers must be placed here, outside most structs that use `HashMap`, I did this to ensure code is clean and readable, I think it is, from my point of view, do you think this should change?.